content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -7,4 +7,57 @@ next: /recipes/authoring/kata-snippets/preloaded/
+
+- **Each requirement should have corresponding fixed test.** Every requirement of the kata described in the description should be explicitly tested by at least one (possibly more) dedicated, properly labeled test case(s).
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. When random test fails, it should be accompanied by at least one failed equivalent fixed test.
+
+## Random Tests
+
+Random tests are not common testing practice and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass. Random tests are necessary especialy for difficult, highly rewarding kata, but Codewars testing guidelines require them for every kata they are applicable for.
+
+- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests.
+- **Order of random tests should be non-deterministic.** Not only inputs should be randomized, but their order too. One possible technique is to generate and collect a set of random inputs for all required scenarios, and shuffle them.
+- If possible, **reference solution should not be used**. Sometimes there's no other way, but quite often reference solution is simply not necessary. Test cases can be generated with the answer known upfront, and it can save time which would be spent on running the reference solution.
+- **Assertions should be close to the calls of tested solution**, preferrably immediatelly following it. Debugging becomes difficult if invalid answer is verified long after it's returned.
+- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
+- **Difficulty should be consistent beetwen test runs.** When randomized ranges are very large, then it's possible that one user receives many small, easy inputs and their solution passes, while another user receives many large, more difficult inputs, and their solution fails or times out. Make sure that your random tests are built in a way which elimnates such situation. If you want to test difficult inputs, split test cases into a set of easy ones and a set of diffficult ones, and tets them separately.
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If possible, Try to compose your test suite in such a way that when test fail, they fail on some small, easy, debuggable input, before they fail on a large one, which is difficult to print. 
```

Comment:
```suggestion
- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. In such a case, two batches of random tests should be provided: a first batch with small debuggable inputs (note that the inputs are small, not the batch. So keep a reasonnable number of tests: around 50 at least), then the batch on wide inputs. The kind of inputs generated in both batches should match, so that the user can actually find the critical cases for his code in the small inputs.
```